src/deluge/gui/ui/load/load_song_ui.cpp

```diff
@@ -227,9 +234,60 @@ ActionResult LoadSongUI::buttonAction(deluge::hid::Button b, bool on, bool inCar
 	return ActionResult::DEALT_WITH;
 }
 
+void LoadSongUI::loadNextSongIfAvailable() {
+	if (currentUIMode == UI_MODE_LOADING_SONG_UNESSENTIAL_SAMPLES_ARMED
+	    || currentUIMode == UI_MODE_LOADING_SONG_UNESSENTIAL_SAMPLES_UNARMED
+	    || currentUIMode == UI_MODE_LOADING_SONG_ESSENTIAL_SAMPLES
+	    || currentUIMode == UI_MODE_LOADING_SONG_NEW_SONG_PLAYING || skipAnimations) {
+		// While in the process of loading a song, don't do anything
+		return;
+	}
+
+	skipAnimations = true; // disable animations while auto-loading a song
+	if (openUI(&loadSongUI)) {
```

Comment:
is it possible to bypass opening the load song UI for this? Just seems weird and I wonder if there's a cleaner way

Reply:
Not sure if possible without replicating code. There are lots of things going on when you open the browser so not sure if the refactoring needed is worth it.

Reply:
yeah I think it would need a split like for synth/kit loading, where the UI can be used without actually being opened. If you take a look at how they handle turning the select encoder to change presets something like that would probably make more sense. My concern is almost purely that I think this will lead to bugs in the future since it's not clear that the load song UI will need to support being used while not being the active UI, and since this is likely a niche use it could probably silently break for a while before being noticed

Reply:
Can you point me in the direction where that split for synth/kit loading is done? Is there any specific PR that refactored the synth browser to be separate? thanks!

Reply:
https://github.com/SynthstromAudible/DelugeFirmware/blob/6f3409c0b87d6f25cf4c775eb7825b09334f7cc9/src/deluge/gui/ui/load/load_instrument_preset_ui.cpp#L1402

This function is used when scrolling presets without having the file browser open. I think it was there originally 

Here it is in action - https://github.com/SynthstromAudible/DelugeFirmware/blob/a5ab97bbb5687514c3e62962a74b263cba7476f0/src/deluge/model/song/song.cpp#L4655 

Something like this is cleaner since it's clear what it does and doesn't have the chance of weird interactions with the UI stack in the future

Reply:
@m-m-adams could you take a look at latest changes? now it doesn't pop the UI

Reply:
I am still getting freezes with loading song from stopped playback state, but I wanted to check if what I have already modified is the direction in which you wanted this PR to go

Reply:
If I touch the grid when the next song is loading due to my new feature I got this once:

0x200e1755: Kit::renderOutput(ModelStack*, StereoSample*, StereoSample*, long, long*, long, long, bool, bool) at :?
0x200e170d: Kit::renderOutput(ModelStack*, StereoSample*, StereoSample*, long, long*, long, long, bool, bool) at /Users/raul/Development/Projects/Deluge/DelugeFirmware/src/deluge/model/instrument/kit.cpp:655
0x2012ff5b: Song::renderAudio(StereoSample*, long, long*, long) [clone .constprop.0] at /Users/raul/Development/Projects/Deluge/DelugeFirmware/src/OSLikeStuff/timers_interrupts/timers_interrupts.h:44
 (inlined by) Song::renderAudio(StereoSample*, long, long*, long) [clone .constprop.0] at /Users/raul/Development/Projects/Deluge/DelugeFirmware/src/deluge/model/song/song.cpp:2421
0x2015a958: RMSFeedbackCompressor::render(StereoSample*, unsigned short, long, long, long) at /Users/raul/Development/Projects/Deluge/DelugeFirmware/src/deluge/dsp/compressor/rms_feedback.cpp:129 (discriminator 1)

why would it crash the deluge? could it be something related to timer interrupts? Maybe the code that loads the next song should be declaring some sort of thread or timer that i am missing?

Reply:
This is another crash, do you find any clue?

0x200e1755: Kit::renderOutput(ModelStack*, StereoSample*, StereoSample*, long, long*, long, long, bool, bool) at :?
0x20133d51: Voice::renderBasicSource(Sound*, ParamManagerForTimeline*, long, long*, long, bool, long, bool*, long, bool, unsigned long*, unsigned long*, long, unsigned long*, bool, long) [clone .constprop.0] at /Users/raul/Development/Projects/Deluge/DelugeFirmware/src/deluge/model/voice/voice.cpp:2463
0x2015b08c: DelayBuffer::isActive() const at /Users/raul/Development/Projects/Deluge/DelugeFirmware/src/deluge/dsp/../dsp/delay/delay_buffer.h:67
 (inlined by) Delay::process(std::span<StereoSample, 4294967295u>, Delay::State const&) at /Users/raul/Development/Projects/Deluge/DelugeFirmware/src/deluge/dsp/delay/delay.cpp:282
0x200e6657: ModControllableAudio::processFX(StereoSample*, long, ModFXType, long, long, Delay::State const&, long*, ParamManager*, bool, long) at /Users/raul/Development/Projects/Deluge/DelugeFirmware/src/deluge/model/mod_controllable/mod_controllable_audio.cpp:177 (discriminator 2)

Reply:
it seems several things depend on detecting if the UI is open:
<img width="760" alt="image" src="https://github.com/user-attachments/assets/e8e7113e-67cb-46ef-aa63-e89f2c65035a">
Is it really necessary to refactor this as part of this PR? seems like overkill :(